modules/standard/MemMove.chpl

```diff
@@ -0,0 +1,412 @@
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Support for move-initializing and deinitializing values.
+
+  The :mod:`MemMove` module provides functions which enable users to
+  move-initialize and deinitialize values.
+
+  The functionality provided by this module can be used to implement
+  collections in a manner similar to those implemented by the Chapel standard
+  modules (such as :mod:`List` or :mod:`Map`).
+
+  .. note::
+
+    Throughout documentation, the term `variable` also includes non-const
+    formals, array elements, record or class fields, and tuple components.
+
+
+*/
+module MemMove {
+
+  // Mark as "unsafe" to silence lifetime errors.
+  pragma "unsafe"
+  private inline proc _move(ref dst, const ref src) {
+    __primitive("=", dst, src);
+  }
+
+  /*
+    Check to see if a given type needs to be deinitialized.
+
+    :arg t: A type to check for deinitialization
+    :type t: `type`
+
+    :return: ``true`` if ``t`` needs to be deinitialized
+    :rtype: param bool
+  */
+  proc needsDeinit(type t) param {
+    return __primitive("needs auto destroy", t);
+  }
+
+  /*
+    Explicitly deinitialize a variable. The variable referred to by ``arg``
+    should be considered uninitialized after a call to this function.
+
+    .. warning::
+
+      At present the compiler does not account for deinitialization performed
+      upon a call to :proc:`explicitDeinit()`. It should only be called
+      when deinitialization would not occur otherwise.
+
+    :arg: A variable to deinitialize
+  */
+  proc explicitDeinit(ref arg: ?t) {
+    if needsDeinit(t) then
+      chpl__autoDestroy(arg);
+  }
+
+  /*
+    Move-initialize ``lhs`` with the value in ``rhs``. The contents of ``lhs``
+    are not deinitialized before the move, and ``rhs`` is not deinitialized
+    after the move.
+
+    .. warning::
+
+      If ``lhs`` references an already initialized variable, it will be
+      overwritten by the contents of ``rhs`` without being deinitialized
+      first. Call :proc:`explicitDeinit()` to deinitialize ``lhs`` if
+      necessary.
+
+    :arg lhs: A variable to move-initialize, whose type matches ``rhs``
+
+    :arg rhs: A value to move-initialize from
+  */
+  pragma "last resort"
+  deprecated "The formals 'lhs' and 'rhs' are deprecated, please use 'dst' and 'src' instead"
+  proc moveInitialize(ref lhs,
```

Comment:
Since this a new module, I'd just avoid adding anything deprecated right away.  Users are already updating their code, might as well avoid the clutter

Reply:
I thought about that, but if they first switch "Memory.Initialization" to "MemMove", then their existing code might not compile and the error messages won't be helpful. This is particularly relevant for the change from "moveToValue" to "moveFrom".

Reply:
They should get the message from both, since deprecations are warnings.  So they should see both messages in their output and if the code doesn't quite work after the first module change, they can look back at their old error message as well.  I think this would fall under the category of making slightly less work for ourselves, but I do agree that the strategy you're using here will do more to ensure they get a good error message throughout.